File: packages/core/src/stores/memory.ts
import type { ChallengeData, ChallengeStore } from '../types.js'

interface Entry {
  data: ChallengeData
  expires_at: number
}

export class MemoryStore implements ChallengeStore {
  private store = new Map<string, Entry>()

  async set(id: string, data: ChallengeData, ttlSeconds: number): Promise<void> {
    this.store.set(id, {
      data,
      expires_at: Date.now() + ttlSeconds * 1000,
    })
  }

  async get(id: string): Promise<ChallengeData | null> {
    const entry = this.store.get(id)
    if (!entry) return null
    if (Date.now() >= entry.expires_at) {
      this.store.delete(id)
      return null
    }
    return entry.data
  }

  async delete(id: string): Promise<void> {
    this.store.delete(id)
  }
}

File: sdks/python/src/xagentauth/stores/memory.py
from __future__ import annotations

import time
from dataclasses import dataclass

from xagentauth.types import ChallengeData


@dataclass
class _Entry:
    data: ChallengeData
    expires_at: float


class MemoryStore:
    """In-memory challenge store with TTL-based expiration."""

    def __init__(self) -> None:
        self._store: dict[str, _Entry] = {}

    async def set(self, id: str, data: ChallengeData, ttl_seconds: int) -> None:
        self._store[id] = _Entry(
            data=data,
            expires_at=time.time() + ttl_seconds,
        )

    async def get(self, id: str) -> ChallengeData | None:
        entry = self._store.get(id)
        if entry is None:
            return None
        if time.time() >= entry.expires_at:
            del self._store[id]
            return None
        return entry.data

    async def delete(self, id: str) -> None:
        self._store.pop(id, None)
